graph/src/data/subgraph/mod.rs

```diff
@@ -666,6 +673,95 @@ impl UnresolvedMapping {
+    }
+
+    pub fn try_from_versions<'a>(
+        versions: impl Iterator<Item = &'a Version>,
+    ) -> Result<Self, DifferentMappingApiVersions> {
+        let unique_versions: BTreeSet<Version> = versions.into_iter().cloned().collect();
+
+        let all_below_referential_version = unique_versions.iter().all(|v| *v < API_VERSION_0_0_5);
+        let all_the_same = unique_versions.len() == 1;
+
+        let unified_version: Option<Version> = match (all_below_referential_version, all_the_same) {
+            (false, false) => return Err(unique_versions.into()),
+            (false, true) => Some(unique_versions.iter().nth(0).unwrap().deref().clone()),
+            (true, _) => None,
+        };
+
+        Ok(UnifiedMappingApiVersion(unified_version))
+    }
+}
+
+#[test]
+fn unified_mapping_api_version_from_iterator() {
+    let input = [
+        vec![Version::new(0, 0, 5), Version::new(0, 0, 5)], // Ok(Some(0.0.5))
+        vec![Version::new(0, 0, 6), Version::new(0, 0, 6)], // Ok(Some(0.0.6))
+        vec![Version::new(0, 0, 3), Version::new(0, 0, 4)], // Ok(None)
+        vec![Version::new(0, 0, 4), Version::new(0, 0, 4)], // Ok(None)
+        vec![Version::new(0, 0, 3), Version::new(0, 0, 5)], // Err({0.0.3, 0.0.5})
+        vec![Version::new(0, 0, 6), Version::new(0, 0, 5)], // Err({0.0.5, 0.0.6})
+    ];
+    let output: [Result<UnifiedMappingApiVersion, DifferentMappingApiVersions>; 6] = [
+        Ok(UnifiedMappingApiVersion(Some(Version::new(0, 0, 5)))),
+        Ok(UnifiedMappingApiVersion(Some(Version::new(0, 0, 6)))),
+        Ok(UnifiedMappingApiVersion(None)),
+        Ok(UnifiedMappingApiVersion(None)),
+        Err(input[4]
+            .iter()
+            .cloned()
+            .collect::<BTreeSet<Version>>()
+            .into()),
+        Err(input[5]
+            .iter()
+            .cloned()
+            .collect::<BTreeSet<Version>>()
+            .into()),
+    ];
+    for (version_vec, expected_unified_version) in input.iter().zip(output.iter()) {
+        let unified = UnifiedMappingApiVersion::try_from_versions(version_vec.iter());
+        match (unified, expected_unified_version) {
+            (Ok(a), Ok(b)) => assert_eq!(a, *b),
+            (Err(a), Err(b)) => assert_eq!(a, *b),
+            _ => panic!(),
+        }
+    }
+}
+
+#[derive(Error, Debug, PartialEq)]
+#[error("Expected a single apiVersion for mapings. Found: {}.", format_versions(.0))]
+pub struct DifferentMappingApiVersions(BTreeSet<Version>);
+
+fn format_versions(versions: &BTreeSet<Version>) -> String {
+    versions.iter().map(ToString::to_string).join(", ")
+}
+
+impl From<BTreeSet<Version>> for DifferentMappingApiVersions {
+    fn from(versions: BTreeSet<Version>) -> Self {
+        Self(versions)
+    }
+}
+
+impl From<DifferentMappingApiVersions> for SubgraphManifestValidationError {
+    fn from(versions: DifferentMappingApiVersions) -> Self {
+        SubgraphManifestValidationError::DifferentApiVersions(versions.0)
+    }
+}
+
```

Comment:
Added this new error so it is ergonomic to handling in other places.

Reply:
Nice!